test/modules/TwinTester/DesiredPropertiesValidator.cs

```diff
@@ -122,24 +114,29 @@ async Task RemovePropertiesFromTwin(Dictionary<string, string> propertiesToRemov
             try
             {
                 string patch = $"{{ properties: {{ desired: {JsonConvert.SerializeObject(propertiesToRemoveFromTwin)} }}";
-                Twin newTwin = await this.registryManager.UpdateTwinAsync(Settings.Current.DeviceId, Settings.Current.TargetModuleId, patch, this.twinState.TwinETag);
-                this.twinState.TwinETag = newTwin.ETag;
+                Twin newTwin = await this.registryManager.UpdateTwinAsync(Settings.Current.DeviceId, Settings.Current.TargetModuleId, patch, this.twinTestState.TwinETag);
+                this.twinTestState.TwinETag = newTwin.ETag;
             }
             catch (Exception e)
             {
                 Logger.LogError(e, "Failed call to remove successful desired property updates.");
             }
         }
 
-        bool ExceedFailureThreshold(TwinState twinState, DateTime twinUpdateTime)
+        bool ExceedFailureThreshold(TwinTestState twinTestState, DateTime twinUpdateTime)
         {
-            DateTime comparisonPoint = twinUpdateTime > twinState.LastTimeOffline ? twinUpdateTime : twinState.LastTimeOffline;
+            DateTime comparisonPoint = twinUpdateTime > twinTestState.LastNetworkOffline ? twinUpdateTime : twinTestState.LastNetworkOffline;
             return DateTime.UtcNow - comparisonPoint > Settings.Current.TwinUpdateFailureThreshold;
         }
 
-        bool ShouldExpectDesiredPropertyCallback(TwinState twinState, DateTime twinUpdateTime)
+        bool SkipCallbackValidationDueToEdgeHubRestart(TwinTestState twinTestState, DateTime twinUpdateTime)
         {
-            return (twinUpdateTime - twinState.LastTimeOfEdgeRestart).Duration() > Settings.Current.EdgeHubRestartFailureTolerance;
+            DateTime edgeHubRestartTolerancePeriodLowerBound =
+                twinTestState.EdgeHubLastStopped == DateTime.MinValue ? DateTime.MinValue : twinTestState.EdgeHubLastStopped.Add(-Settings.Current.EdgeHubRestartFailureTolerance);
+            DateTime edgeHubRestartTolerancePeriodUpperBound =
+                twinTestState.EdgeHubLastStarted == DateTime.MinValue ? DateTime.MinValue : twinTestState.EdgeHubLastStarted.Add(Settings.Current.EdgeHubRestartFailureTolerance);
+
+            return edgeHubRestartTolerancePeriodLowerBound <= twinUpdateTime && twinUpdateTime <= edgeHubRestartTolerancePeriodUpperBound;
```

Comment:
We will also need to apply this failure tolerance when operations to ModuleClient.GetTwin() fail. We shouldn't report these errors if caused by a long edgeHub restart.

This will logic change will affect making reported property updates and the call to GetTwin() in this DesiredPropertiesValidator.cs (line 37).

Reply:
Talked to Andrew that we should agree to wrap module client with edgehub restart tolerance logic.  But we also agreed to do it in different PR.